### PR TITLE
Incorporated Best Practices

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
TDL-13636: Best Practices
- Updated singer-python version from 5.3.3 -> 5.12.1
- Updated pendulum version from 1.2.0 -> 2.1.2
- Changed sandbox -> tap_tester_sandbox
- tap-tester image is up-to-date
- Added all integration test
    - Test Sync Canary
    - Test Discovery Mode
    - Test Start Date
    - Test Pagination (for "products" stream)

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Review by dbshah1212 — CHANGES_REQUESTED** [review]

File: tests/base.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+            'unsubscribers': {
+                self.PRIMARY_KEYS: {'id'}
+            }
+        }
+
+    # def expected_check_streams(self):
+    #     return {
+    #         'products',
+    #         'reviews'
+    #     }
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/base.py]: Done

File: tests/base.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    def expected_streams(self):
+        """A set of expected stream names"""
+        return set(self.expected_metadata().keys())
+
+    # def expected_pks(self):
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/base.py]: Done

File: tests/test_yotpo_start_date.py [hunk 3]
```diff
@@ -106,11 +131,11 @@ def test_run(self):
 
                     # Verify start_date key values are greater than or equal to start date of sync 1
                     for start_date_key_value in start_date_key_sync_1:
-                        self.assertGreaterEqual(self.dt_to_ts(start_date_key_value), start_date_1_epoch)
+                        self.assertGreaterEqual(self.dt_to_ts(start_date_key_value), (start_date_1_epoch - 30*24*3600) if reduce_epoch else start_date_1_epoch)
```
Comment (dbshah1212) [inline tests/test_yotpo_start_date.py:134]: Store 30*24*3600 in one variable.

**Review by dbshah1212 — CHANGES_REQUESTED** [review]

File: tests/test_yotpo_start_date.py (hunk 3 above)
Comment (harshpatel4crest) [reply tests/test_yotpo_start_date.py:134]: Okay

**Review by KAllan357 — APPROVED** [review]

**Pull request merged**